Allow cluster access during node restart

henningandersen opened the pull request:
This commit modifies InternalTestCluster to allow using client() and
other operations inside a RestartCallback (onStoppedNode typically).
Restarting nodes are now removed from the map and thus all
methods now return the state as if the restarting node does not exist.

This avoids various exceptions stemming from accessing the stopped
node(s).

Part of #42518

henningandersen committed:
Allow cluster access during node restart

This commit modifies InternalTestCluster to allow using client() and
other operations inside a RestartCallback (onStoppedNode typically). It
goes in a direction of most methods returning the state as if the
restarting node did not exist, avoiding various exceptions stemming from
accessing the stopped node(s).

elasticmachine commented:
Pinging @elastic/es-distributed

henningandersen committed:
getInstance(clazz, node-name) support closed

When explicitly asking for an instance from a node, return it also from
closed nodes.

ywelsch commented on test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java:
```diff
@@ -1086,16 +1095,18 @@ private synchronized void reset(boolean wipeData) throws IOException {
-            expectedNodes.add(getInstanceFromNode(ClusterService.class, nodeAndClient.node()).localNode());
-        }
+        Set<DiscoveryNode> expectedNodes =
+            nodes.values().stream()
```
should we temporarily remove the node from the `nodes` map when it is restarted during the close? This could possibly simplify things here, not requiring us to exclude closed nodes everywhere. WDYT?

henningandersen replied:
I originally did not do so due to test failures. In hindsight, I think I agree, it will leave this in a much cleaner state. Will give it a go and see if I can fix the tests that needs fixing (at least GatewayIndexStateIT.testArchiveBrokenClusterSettings.

ywelsch reviewed (changes requested):
Added a suggestion on how we could possibly do this in a simpler way

henningandersen committed:
Merge remote-tracking branch 'origin/master' into enhance_test_framework_disregard_nodes_during_restart

henningandersen committed:
Restarting nodes unavailable

Changed strategy to simply remove the nodes from the nodes map during
restart. This simplifies this and cements that you cannot get a hold of
a restarting node in onStopped. Fixed the 3 violating cases to comply
with that.

henningandersen committed:
Fixed rebuild of unicast hosts during restart.

henningandersen committed:
Further simplification of restartNode.

henningandersen committed:
Whitespace issue.

henningandersen commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into enhance_test_framework_disregard_nodes_during_restart

henningandersen commented:
Thanks for reviewing @ywelsch , this is now much simpler and ready for another round.

ywelsch commented on test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java:
```diff
@@ -1648,20 +1648,9 @@ private void restartNode(NodeAndClient nodeAndClient, RestartCallback callback)
-            }
-        }
-
-        if (activeDisruptionScheme != null) {
```
why remove this? Is it a NOOP?

henningandersen replied:
We now call publishNode, which both adds to the nodes map and applies active disruption scheme.

ywelsch reviewed (approved):
LGTM
